4_evaluate/src/interpolate_predictions.py

```diff
@@ -0,0 +1,225 @@
+            # np.nanargmin will error
+            no_nans = ~np.isnan(np.nanmin(start_date_matrix, axis=0))
+            which_sequence = np.nanargmin(start_date_matrix[:, no_nans], axis=0)
+            sequence_indices_batch = np.array([-1]*this_batch_size)
+            sequence_indices_batch[no_nans] = idx_sequence[which_sequence]
+            # Now, get the index of the date within the sequence for each obs
+            date_indices_batch = np.array([-1]*this_batch_size)
+            date_indices_batch[no_nans] = (obs_dates_days[batch_start:batch_end][no_nans] - 
+                                           pred_starts_days[sequence_indices_batch[no_nans]])
+            sequence_indices[batch_start:batch_end] = sequence_indices_batch
+            date_indices[batch_start:batch_end] = date_indices_batch
+
+    # Now, get the prediction depth above and below each observation
+    obs_depths = obs.depth.to_numpy()
+    # Clip in order to extrapolate outside predicted depth range
+    clipped_obs_depths = np.clip(obs_depths, min(depths_use), max(depths_use))
+    # Pad depths with one value that is always below obs,
+    # for convenience in interpolation algorithm below
+    pad_value = max(np.r_[depths_use, clipped_obs_depths] + 1)
+    padded_pred_depths = np.r_[depths_use, pad_value]
+
+    # Get index and value of next largest depth in list of depths
+    def enumerate_depth_below(depth):
+        return next((ii, x) for ii, x in enumerate(padded_pred_depths) if x > depth)
+    depth_below_tuples = [enumerate_depth_below(obs_depth) for obs_depth in clipped_obs_depths]
+
+    # Unpack list of two-element tuples into two arrays
+    depth_below_index, depth_below = (np.array(tp) for tp in zip(*depth_below_tuples))
+    depth_above_index = depth_below_index - 1
+    depth_above = depths_use[depth_above_index]
+
+    # Compute interpolation factors
+    above_factors = (depth_below - clipped_obs_depths)/(depth_below - depth_above)
+    below_factors = (clipped_obs_depths - depth_above)/(depth_below - depth_above)
+    # Check that the factors sum to 1
+    assert all(above_factors + below_factors == 1)
+
+    # Now, some depth_below_index values are one greater than the maximum index in depths_use
+    # The corresponding below_factors should be zero, though
+    # Let's double check that
+    too_low = np.nonzero(depth_below_index==len(depths_use))
+    assert all(below_factors[too_low] == 0)
+    # Good. That means that the depth_below_index doesn't need to be used when it's one greater than max
+    # because those values get multiplied by zero
+    # Set those indices to the max index instead
+    depth_below_index[too_low] = len(depths_use) - 1
+
+    # Let's now skip any obs with indices equal to -1
+    # TODO: Consider dropping observations that are deeper than deepest depth_use
+    use_obs = ((sequence_indices >= 0) & (date_indices >= 0))
+
+    # Now the tricky part (like it hasn't been tricky so far!):
```

Comment:
Yes 😅. This looks like it was a lot of work that is tricky to think-up from scratch

Reply:
Haha, maybe too tricky... but it's a working solution!